### PR TITLE
Shrink `TableConfig` in `StateMachine` tests

### PR DESCRIPTION
**Pull request by jorisdral**
The `TableConfig` arbitrary instance is still quite suboptimal. I've added a TODO to improve it.

### CODE REVIEW

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 1]
```diff
@@ -192,17 +192,58 @@ propLockstep_ModelIOImpl =
...
 
+  shrink R.TableConfig{..} =
+      [ R.TableConfig {
+            confWriteBufferAlloc = confWriteBufferAlloc'
+          , ..
+          }
+      | confWriteBufferAlloc' <- QC.shrink confWriteBufferAlloc
+      ]
+
+-- TODO: the current generator is suboptimal, and should be improved. There are
```
Comment (dcoutts) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:215]:
```suggestion
-- | TODO: the current generator is suboptimal, and should be improved. There are
```

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 2]
```diff
@@ -192,17 +192,58 @@ propLockstep_ModelIOImpl =
...
+--   cover in the state machine tests.
+--
+-- * If @n@ is too small, then we flush/merge very frequently, which may exhaust
+--   the number of open file handles when using the real file system. Specially
+--   if the test fails and the shrinker kicks in, then @n@ can currently become
+--   very small. This is good for finding a minimal counterexample, but it might
+--   also make the real filesystem run out of file descriptors. Arguably, you
+--   could overcome this specific issue by only generating or shrinking to small
+--   @n@ when we use the mocked file system. This would require some boilerplate
+--   to add type level tags to distinguish between the two cases.
```
Comment (dcoutts) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:236]: This is all a good explanation of the problem, and picking (1,30] seems like quite a good interim solution.

I would be surprised if we did run out of file handles. The number of file handles per table is still log base 4, so even with size 1 write buffer, a tree with 6 levels is already enormous, so that's 6 * 8 = 48. And ok, there can be many tables, but there's also a lot of sharing.

I think it's also reasonable to run the IO tests with more file descriptors available. The 1024 is a completely artificial default "soft" limit. One can just raise it using `ulimit -S -n 4096` or whatever before running the test. We can do that in CI.

**Review by dcoutts — APPROVED** [review]
LGTM, and good explanation of the WB size issue.

**Comment by dcoutts** [comment]
Perhaps we should by default go for these small WB sizes so we can get good coverage of levels and flushing, merging etc. And then separately have some more targeted tests for large things with big values, overflow pages etc.

File: test/Test/Database/LSMTree/Normal/StateMachine.hs (hunk 2 above)
Comment (jorisdral) [reply test/Test/Database/LSMTree/Normal/StateMachine.hs:236]: The test has run out of file descriptors for me, but only if I up the number of tests and the maximum size a bit. It's still relatively rare, but it would be best if we did not have such a flaky test

**Comment by jorisdral** [comment]
> Perhaps we should by default go for these small WB sizes so we can get good coverage of levels and flushing, merging etc. And then separately have some more targeted tests for large things with big values, overflow pages etc.

I've been thinking the same thing. We could probably write our own `DL` expressions that test large write buffer cases
